src/upgrades/mongodb_upgrade.py

```diff
@@ -191,28 +245,60 @@ def _set_upgrade_status(self):
 
         # Set/clear upgrade unit status if no other unit status - upgrade status for units should
         # have the lowest priority.
-        if isinstance(self.charm.unit.status, ActiveStatus):
+        if isinstance(self.charm.unit.status, ActiveStatus) or (
+            isinstance(self.charm.unit.status, BlockedStatus)
+            and self.charm.unit.status.message.startswith(
+                "Rollback with `juju refresh`. Pre-upgrade check failed:"
+            )
+        ):
             self.charm.unit.status = self._upgrade.get_unit_juju_status() or ActiveStatus()
 
+    def wait_for_cluster_healthy(self) -> None:
+        """Waits until the cluster is healthy after upgrading.
+
+        After a unit restarts it can take some time for the cluster to settle.
+
+        Raises:
+            ClusterNotHealthyError.
+        """
+        for attempt in Retrying(stop=stop_after_attempt(10), wait=wait_fixed(1)):
+            with attempt:
+                if not self.is_cluster_healthy():
+                    raise ClusterNotHealthyError()
+
     def is_cluster_healthy(self) -> bool:
         """Returns True if all nodes in the cluster/replcia set are healthy."""
-        if self.charm.is_role(Config.Role.SHARD):
-            logger.debug("Cannot run full cluster health check on shards")
-            # TODO Future PR - implement cgecj healthy check for single shard
-            return False
+        with MongoDBConnection(
+            self.charm.mongodb_config, "localhost", direct=True
+        ) as direct_mongo:
+            if not direct_mongo.is_ready:
+                logger.error("Cannot proceed with upgrade. Service mongod is not running")
+                return False
 
         try:
-            if self.charm.is_role(Config.Role.CONFIG_SERVER):
-                # TODO Future PR - implement node healthy check for sharded cluster
+            if self.charm.is_role(Config.Role.CONFIG_SERVER) or self.charm.is_role(
+                Config.Role.SHARD
+            ):
+                # TODO Future PR - implement node healthy check for entire cluster
                 return False
             if self.charm.is_role(Config.Role.REPLICATION):
-                return self.are_replica_set_nodes_healthy(self.charm.mongodb_config)
+                node_healthy = self.are_replica_set_nodes_healthy(self.charm.mongodb_config)
```

Comment:
missing a return

Reply:
no, the return is performed below, but I think this can be cleaned up / reorganised. I'll do that